test/general/test_items.py

```diff
@@ -163,3 +163,84 @@ def test_locality_not_modified(self):
                                          f"{game_name} modified local_items during {step}")
                         self.assertEqual(non_local_items, multiworld.worlds[1].options.non_local_items.value,
                                          f"{game_name} modified non_local_items during {step}")
+
+    def test_collect_remove_location_accessibility(self):
+        """
+        Test that worlds' .remove() implements the reverse of .collect() by checking location accessibility, and test
+        that collecting an item never reduces accessibility.
+
+        Because logic issues could be present only when items are collected in a specific order, there is no guaranteed
+        way to identify all collect/remove issues, so the test collects items in a deterministically random order.
+
+        test_create_item only tests that .collect()/.remove() correctly update CollectionState.prog_items, but worlds
+        can implement their own data structures on CollectionState objects. These custom data structures being updated
+        correctly by .collect()/.remove() are tested by comparing location accessibility before and after
+        .collect() + .remove().
+        """
+        for game_name, world_type in AutoWorldRegister.world_types.items():
+            multiworld = setup_solo_multiworld(world_type)
+            with self.subTest(game=game_name, seed=multiworld.seed):
+                # Get all the advancement items in the multiworld.
+                items = [item for item in multiworld.get_items() if item.advancement]
+
+                # Get all locations in the multiworld.
+                locations = multiworld.get_locations()
+
+                # Shuffle the items into a deterministically random order because otherwise, the order of all items
+                # should always be the same, limiting what can be tested.
+                multiworld.random.shuffle(items)
```

Comment:
My main concern about this test is that, while worlds must account for items in the datapackage being collected in any order and in any amounts, this same requirement might not necessarily be the case for pre-placed event items.

Perhaps there could be a world that knows that certain events will always be collected in a specific order and has logic built around this, but does not account for something like this test trying to collect the events in an otherwise impossible order, which could result in accessibility being reduced when collecting the events out-of-order and therefore failing this test.

Reply:
A have a working version that sweeps for all reachable events, like `MultiWorld.get_sendable_spheres()`, after each collected non-event item, which results in events being collected in a natural order. So I could update the PR with this change.